Uplift azure

aiyengar2 opened the pull request:
This PR updates github.com/Azure/azure-sdk-for-go to 39.1.0 and github.com/Azure/go-autorest to 13.3.3.

Major updates:
- Redefines AuthenticateDeviceFlow and AuthenticateServicePrincipal (now renamed AuthenticateClientCredentials) as functions that return Authorizer objects after running ValidateAuthorizer instead of tokens
- Simplifies / eliminates several helper functions to use newer autorest libraries for Azure authentication
- Introduced arm.go + customimage.go to create a unified way of referring to resources that
are parsed from ARM resource identifiers
- Using contexts to make calls via the Azure client and evaluating the status of async requests using Azure futures (which support WaitForCompletionRef and future.Result), which removes redundant Get calls
- Node driver now implements Kill() as Azure now supports a force kill of an instance
- Remove isSwarm logic as docker swarm is no longer supported

Misc. Updates:
- rename Azure Driver.ctx to Driver.deploymentContext to avoid ambiguity with Go contexts
- Registered Microsoft.Subscription and Microsoft.Resources as resource providers
- Added some more necessary client logic to clients.go for future commits
- moved some Azure fmts to naming.go to centralize naming logic in one file
- Added more structs for different types of Azure resources to cleanup.go to create one common way of cleaning up resources instead of having redundant code
- Changes supportedEnvironments from a map to a list and adds azure.EnvironmentFromName to get the corresponding environment instead.
- Renamed HasAttachedResources to CanBeDeleted and flipped boolean logic accordingly

Related Issue: https://github.com/rancher/rancher/issues/25342

aiyengar2 commented:
Apologies for the large PR; I split this change into 4 separate commits with verbose commit messages to make it easier to read, but a broad majority of the file changes are due to changes in vendor/ from dependency upgrades

luthermonson commented on drivers/azure/azureutil/customimage.go:
```diff
@@ -0,0 +1,117 @@
+			ltrimmed[i] = prefixRemoved[1]
+		} else if i < numExpectedRequiredValues {
+			// failed to parse required values
+			log.Info("image parsed", l)
```
probably dont need

aiyengar2 replied:
removed log in latest commit

aiyengar2 commented:
Squashed 4 commits to 2 which separate vendor changes from code changes

aiyengar2 commented:
@superseb yes with the following exceptions:
- Calling `Kill()` in the previous version of the driver would log `Azure does not implement kill. Calling Stop instead.` and proceed to stop a virtual machine via an Azure VM `PowerOff` API call with no flag for a force stop. In the new driver, `Kill()` directly calls the Azure VM `PowerOff` API and passes in a flag for a force stop to Azure.
- The old driver still responds to opening up Docker Swarm ports based on `d.BaseDriver.SwarmMaster` here, which the new driver does not support: https://github.com/rancher/machine/blob/8cfc2ad0984efc04b41bc32f84d3e2791e8bad1b/drivers/azure/util.go#L120-L135
- If a user to submitted an invalid FaultDomainCount value (i.e. the number of availability zones that you want fault domains in for your VM) or an OS disk name that references an object that exists in another region, the old driver would fail on launch; the new driver fails on PreCreate.
- The old driver failed if a user supplied an image name that was not in the format publisher:offer:sku:version; the new driver allows both the original format as well as a custom image format (i.e. any string that contains an ARM resource identifier for the object)

Other than the above exceptions, all changes are just updating the underlying Azure APIs, so the behavior for creating a node with the current driver should have the same outcome as creating/replacing a node with this driver.

superseb commented on drivers/azure/azureutil/customimage.go:
```diff
@@ -0,0 +1,116 @@
+		} else if i < numExpectedRequiredValues {
+			// failed to parse required values
+			if i == 3 {
+				log.Warn(versionNotSuppliedWarnString)
```
If this is required when using gallery, can we make it part of the error? The error should contain enough to root cause the issue, and while the generic error probably covers it, we do know what the error is but now its only logged in a different level and on a separate line. If it's not required/breaking then this can be the additional warning

aiyengar2 replied:
Good idea! Should be fixed in the latest commit

superseb commented on drivers/azure/azure.go:
```diff
@@ -334,33 +335,65 @@ func (d *Driver) PreCreateCheck() (err error) {
+		if resp.Err != nil {
+			return err
+		} else if d.FaultCount > resp.MaximumFaultDomainCount {
+			return fmt.Errorf("location %s has MaximumPlatformFaultDomainCount=%d, recieved FaultDomainCount=%d",
```
typo received

aiyengar2 replied:
Thanks for catching this! Fixed in the latest commit.

superseb commented:
Can we double check there are no fixes we want in later releases of vendored libraries? And can we also check the API versions used are generally available

aiyengar2 commented:
FYI in my latest commit, I ported a couple of changes / fixes relevant to this PR from https://github.com/rancher/machine/pull/77 since it seems like they make more sense as part of this PR (i.e. uplifting Azure) rather than the NSG PR:
- Added `arm.go`: this file will be used to centralize all ARM resource identifier logic as the same logic is used for NSGs in the commit mentioned above as well as for custom images in this commit.
- Modified `customimages.go`: now uses the logic within `arm.go`
- Modified `cleanup.go`: Renamed HasAttachedResources to CanBeDeleted and flipped boolean logic accordingly in all cleanup functions
- Modified `azureutil.go`: misc. bugs / rewords in `RegisterResourceProviders` + `OSDiskExists` + `virtualNetworkExists` + `CreateVirtualMachine` + `CreateAvailabilitySetIfNotExists`
- Modified `sku.go`: misc. bug in `GetMaximumFaultDomainCount`

aiyengar2 commented:
> Can we double check there are no fixes we want in later releases of vendored libraries? And can we also check the API versions used are generally available

Good call! As I was double checking I discovered that `github.com/Azure/azure-sdk-for-go/storage` will be deprecated soon and realized that we could just use the other storage API now to make all the relevant calls. I've updated the PR accordingly in the latest commit.

Also I just checked / updated the following vendored libraries:
1. `github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-12-01/compute`
1. `github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-12-01/network`
1. `github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources`
1. `github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-11-01/subscriptions`
1. `github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-06-01/storage`

Afaik there are no more pending fixes that we require from these libraries for machine as they should be in their latest versions. Azure puts any service libraries that are on public preview under `github.com/Azure/azure-sdk-for-go/services/preview/<pkg>`, so the versioned APIs should also be GA.

aiyengar2 commented:
This PR will resolve the following issues:
- https://github.com/rancher/rancher/issues/23350
- https://github.com/rancher/rancher/issues/25342 (only milestone 1 + 2, SLES and NSG fixes are part of another PR)
- https://github.com/rancher/rancher/issues/23010

aiyengar2 commented:
> This should only be using the go.mod files, not the gopkg stuff.

fixed in the latest commit!

aiyengar2 commented:
As requested offline, this PR has been split into two separate PRs. This PR now only focuses on the Azure client upliftment whereas the PR linked above adds in custom images / gallery image support.

dramich commented on drivers/azure/azureutil/timeouts.go:
```diff
@@ -0,0 +1,25 @@
+package azureutil
```
The timeouts should be moved to where they are needed, one is even exported as it's used in a different package. It also looks like some of these are not used. If we don't need them they can be deleted.

dramich replied:
And can these be set somehow? If they can't be they should be changed to a const.

aiyengar2 replied:
Yeah that makes sense. Moving them to their respective files in the latest commit.

They shouldn't be set so I'll change them to a const too.

luthermonson commented on drivers/azure/azure.go:
```diff
@@ -334,33 +335,65 @@ func (d *Driver) PreCreateCheck() (err error) {
+		log.Warn("Timed out trying to parse the MaximumFaultDomainCount. Continuing execution...")
+	case resp := <-r:
+		if resp.Err != nil {
+			return err
```
shouldnt this return rest.Err? err at this point is going to be from the new client call above.

aiyengar2 replied:
ah good catch. fixed in the latest commit!
